File: android/PlayKitOVPStarter/app/src/main/java/io/streamroot/dna/samples/playkit/ovpstarter/PlayKitQoSModule.java
package io.streamroot.dna.samples.playkit.ovpstarter;

import com.kaltura.playkit.Player;
import com.kaltura.playkit.PlayerEvent;

import io.streamroot.dna.core.PlaybackState;
import io.streamroot.dna.core.QosModule;

final class PlayKitQoSModule extends QosModule {
    PlayKitQoSModule(Player player) {
        super();

        player.addListener(this, PlayerEvent.stateChanged, event -> {
            switch (event.newState) {
                case IDLE:
                    playbackStateChange(PlaybackState.IDLE);
                case BUFFERING:
                    playbackStateChange(PlaybackState.BUFFERING);
                default:
                    break;
            }
        });

        player.addListener(this, PlayerEvent.pause, pkEvent -> playbackStateChange(PlaybackState.PAUSING));
        player.addListener(this, PlayerEvent.playing, pkEvent -> playbackStateChange(PlaybackState.PLAYING));
        player.addListener(this, PlayerEvent.seeked, pkEvent -> playbackStateChange(PlaybackState.SEEKING));
        player.addListener(this, PlayerEvent.ended, pkEvent -> playbackStateChange(PlaybackState.ENDED));
        player.addListener(this, PlayerEvent.error, pkEvent -> playbackErrorOccurred());
        player.addListener(this, PlayerEvent.videoTrackChanged, pkEvent -> trackSwitchOccurred());
        player.addListener(this, PlayerEvent.videoFramesDropped, pkEvent -> droppedFrameOccurred((int) pkEvent.droppedVideoFrames, pkEvent.droppedVideoFramesPeriod));
    }
}

File: ios/PhoenixAnalyticsSample/PhoenixAnalyticsSample-Swift/PlayKitQoSModule.swift
//
//  PlayKitQoSModule.swift
//  PlayKitOVPStarter
//
//  Created by Jean-Baptiste Louazel on 13/06/2019.
//  Copyright © 2019 Kaltura. All rights reserved.
//

import PlayKit
import StreamrootSDK

// Streamroot QoS module
// Binds player' events to raise precise data to DNAClient
class PlayKitQoSModule {
    let dnaQoSModule: QosModule = DNAQos.module(type: .custom)

    public init(_ player: Player) {
        player.addObserver(self, event: PlayerEvent.stateChanged, block: { event in
            switch (event.newState) {
            case .idle:
                self.dnaQoSModule.playerStateDidChange(PlaybackState.idle)
            case .ready:
                player.isPlaying
                    ? self.dnaQoSModule.playerStateDidChange(PlaybackState.playing)
                    : self.dnaQoSModule.playerStateDidChange(PlaybackState.paused)
            case .buffering:
                self.dnaQoSModule.playerStateDidChange(PlaybackState.buffering)
            case .ended:
                self.dnaQoSModule.playerStateDidChange(PlaybackState.ended)
            case .error:
                self.dnaQoSModule.playbackErrorOccurred()
            default: break
            }
        })

        player.addObserver(self, event: PlayerEvent.videoTrackChanged, block: { _ in self.dnaQoSModule.trackSwitchOccurred() })

        player.addObserver(self, event: PlayerEvent.seeked, block: { _ in self.dnaQoSModule.playerStateDidChange(PlaybackState.seeking) })
        
        player.addObserver(self, event: PlayerEvent.playbackStalled, block: { _ in self.dnaQoSModule.playbackErrorOccurred() })
        
        player.addObserver(self, event: PlayerEvent.ended, block: { _ in self.dnaQoSModule.playerStateDidChange(PlaybackState.ended) })
    }
}
